Handle `macro_rules!` tokens consistently across crates

Aaron1011 opened the pull request:
When we serialize a `macro_rules!` macro, we used a 'lowered' `TokenStream` for its body, which has all `Nonterminal`s expanded in-place via `nt_to_tokenstream`. This matters when an 'outer' `macro_rules!` macro expands to an 'inner' `macro_rules!` macro - the inner macro may use tokens captured from the 'outer' macro in its definition.

This means that invoking a foreign `macro_rules!` macro may use a different body `TokenStream` than when the same `macro_rules!` macro is invoked in the same crate. This difference is observable by proc-macros invoked by a `macro_rules!` macro - a `None`-delimited group will be seen in the same-crate case (inserted when convering `Nonterminal`s to the `proc_macro` crate's structs), but no `None`-delimited group in the cross-crate case.

To fix this inconsistency, we now insert `None`-delimited groups when 'lowering' a `Nonterminal` `macro_rules!` body, just as we do in `proc_macro_server`. Additionally, we no longer print extra spaces for `None`-delimited groups - as far as pretty-printing is concerned, they don't exist (only their contents do). This ensures that `Display` output of a `TokenStream` does not depend on which crate a `macro_rules!` macro was invoked from.

This PR is necessary in order to patch the `solana-genesis-programs` for the upcoming hygiene serialization breakage (https://github.com/rust-lang/rust/pull/72121#issuecomment-646924847). The `solana-genesis-programs` crate will need to use a proc macro to re-span certain tokens in a nested `macro_rules!`, which requires us to consistently use a `None`-delimited group.

See `src/test/ui/proc-macro/nested-macro-rules.rs` for an example of the kind of nested `macro_rules!` affected by this crate.

Aaron1011 committed:
Insert NoDelim groups around nonterminals when lowering macro_rules

rust-highfive commented:
r? @eddyb

(rust_highfive has picked a reviewer for you, use r? to override)

Aaron1011 commented:
r? @petrochenkov

rust-highfive commented:
The job `x86_64-gnu-llvm-8` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=32521&view=logs&jobId=698b0e4a-0480-533d-81fa-f38599aa328e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32521/logs/69), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32521/logs/69)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux x86_64-gnu-llvm-8
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 4'
Agent machine name: 'fv-az578'
Current agent version: '2.170.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200614.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200614.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.163.3)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/0ad4fc1e-4e60-43d5-84d5-4bb087a8a7a2.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73569/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73569/merge:refs/remotes/pull/73569/merge
---
 ---> 31fea614d2f3
Step 5/8 : ENV RUST_CONFIGURE_ARGS       --build=x86_64-unknown-linux-gnu       --llvm-root=/usr/lib/llvm-8       --enable-llvm-link-shared       --set rust.thin-lto-import-instr-limit=10
 ---> Using cache
 ---> 4195cadf126d
Step 6/8 : ENV SCRIPT python2.7 ../x.py test --exclude src/tools/tidy &&            python2.7 ../x.py test src/test/mir-opt --pass=build                                   --target=armv5te-unknown-linux-gnueabi &&            python2.7 ../x.py test src/tools/tidy
 ---> 4e90f6b48f05
Step 7/8 : ENV NO_DEBUG_ASSERTIONS=1
 ---> Using cache
 ---> dfa0a356d899
---
Set({"src/librustc_parse_format"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_passes"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_plugin_impl"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_privacy"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_query_system"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_save_analysis"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_serialize"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_session"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_span"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
---
Set({"src/librustc_parse_format"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_passes"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_plugin_impl"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_privacy"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_query_system"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_save_analysis"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_serialize"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_session"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_span"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
---
   Compiling rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
   Compiling chalk-rust-ir v0.10.0
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling chalk-solve v0.10.0
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
   Compiling rustc_arena v0.0.0 (/checkout/src/librustc_arena)
   Compiling rustc_macros v0.1.0 (/checkout/src/librustc_macros)
   Compiling chalk-derive v0.10.0
   Compiling rustc_span v0.0.0 (/checkout/src/librustc_span)
   Compiling chalk-ir v0.10.0
thread 'rustc' panicked at 'rustc_layout_scalar_valid_range takes args', src/librustc_middle/ty/context.rs:1032:25

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: ***/blob/master/CONTRIBUTING.md#bug-reports
note: rustc 1.46.0-nightly (6d6307ce8 2020-06-21) running on x86_64-unknown-linux-gnu


note: compiler flags: -Z macro-backtrace -Z unstable-options -Z binary-dep-depinfo -Z force-unstable-if-unmarked -C opt-level=3 -C embed-bitcode=no -C debuginfo=0 -C link-args=-Wl,-rpath,$ORIGIN/../lib -C prefer-dynamic -C llvm-args=-import-instr-limit=10 -C debug-assertions=n --crate-type lib
note: some of the compiler flags provided by cargo are hidden

error: could not compile `rustc_span`.

---
  local time: Sun Jun 21 02:23:57 UTC 2020
  network time: Sun, 21 Jun 2020 02:23:57 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73569/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73569/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3588) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-8` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=32528&view=logs&jobId=698b0e4a-0480-533d-81fa-f38599aa328e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32528/logs/69), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32528/logs/69)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux x86_64-gnu-llvm-8
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 61'
Agent machine name: 'fv-az578'
Current agent version: '2.170.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200614.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200614.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.163.3)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/6302fa7a-8abc-4ea1-8c4a-bb48552d806d.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73569/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73569/merge:refs/remotes/pull/73569/merge
---
 ---> 31fea614d2f3
Step 5/8 : ENV RUST_CONFIGURE_ARGS       --build=x86_64-unknown-linux-gnu       --llvm-root=/usr/lib/llvm-8       --enable-llvm-link-shared       --set rust.thin-lto-import-instr-limit=10
 ---> Using cache
 ---> 4195cadf126d
Step 6/8 : ENV SCRIPT python2.7 ../x.py test --exclude src/tools/tidy &&            python2.7 ../x.py test src/test/mir-opt --pass=build                                   --target=armv5te-unknown-linux-gnueabi &&            python2.7 ../x.py test src/tools/tidy
 ---> 4e90f6b48f05
Step 7/8 : ENV NO_DEBUG_ASSERTIONS=1
 ---> Using cache
 ---> dfa0a356d899
---
Set({"src/librustc_parse_format"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_passes"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_plugin_impl"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_privacy"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_query_system"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_save_analysis"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_serialize"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_session"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_span"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
---
Set({"src/librustc_parse_format"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_passes"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_plugin_impl"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_privacy"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_query_system"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_save_analysis"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_serialize"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_session"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_span"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
---
   Compiling rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
   Compiling chalk-rust-ir v0.10.0
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling chalk-solve v0.10.0
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
   Compiling rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
   Compiling chalk-rust-ir v0.10.0
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling chalk-solve v0.10.0
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
...............i.................................................................................... 1900/10376
.................................................................................................... 2000/10376
.........................................i..i....................................................... 2100/10376
.................................................................................................... 2200/10376
...............................iiiii................................................................ 2300/10376
.................................................................................................... 2500/10376
.................................................................................................... 2600/10376
.................................................................................................... 2700/10376
.................................................................................................... 2800/10376
---
.................................................................................................... 5300/10376
.................................................................................................... 5400/10376
...................i................................................................................ 5500/10376
.............i...................................................................................... 5600/10376
.................................ii.ii........i...i................................................. 5700/10376
.i...............................................................................................i.. 5900/10376
.................................................................................................... 6000/10376
...................................................ii.....................................i......... 6100/10376
.................................................................................................... 6200/10376
.................................................................................................... 6200/10376
.................................................................................................... 6300/10376
.................................................................................................... 6400/10376
..............ii...i..ii...........i................................................................ 6500/10376
.................................................................................................... 6700/10376
.................................................................................................... 6800/10376
.................................................................................................... 6800/10376
................................................i..ii............................................... 6900/10376
.................................................................................................... 7100/10376
.................................................................................................... 7200/10376
....i............................................................................................... 7300/10376
.................................................................................................... 7400/10376
---
.................................................................................................... 8300/10376
.................................................................................................... 8400/10376
..................................................i................................................. 8500/10376
.................................................................................................... 8600/10376
....iiiiii..iiiiii.i................................................................................ 8700/10376
.................................................................................................... 8900/10376
.................................................................................................... 9000/10376
.................................................................................................... 9100/10376
.................................................................................................... 9200/10376
---
Suite("src/test/codegen") not skipped for "bootstrap::test::Codegen" -- not in ["src/tools/tidy"]
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 197 tests
iiii......i................ii.i..........i......................i...........i..i........i........i.. 100/197
..i.............i.i.i...iii..iiii....................................iii.................ii......

 finished in 6.664
Suite("src/test/codegen-units") not skipped for "bootstrap::test::CodegenUnits" -- not in ["src/tools/tidy"]
Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 22 tests
iiiiiiiiiiiiiiiiiiiiii

 finished in 0.162
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiii..i.....i..i...i..i.i.i..i..i..ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii. 100/116
....iiii.....ii.

 finished in 15.595
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
.................................................................................................... 100/1026
.................................................................................................... 200/1026
.................................................................................................... 300/1026
.................................................................................................... 400/1026
..FFF....F...............FF.F..........ii........................................................... 500/1026
.................................................................................................... 700/1026
.................................................................................................... 800/1026
.................................................................................................... 900/1026
.................................................................................................... 1000/1026
.................................................................................................... 1000/1026
..........................
failures:

---- num::dec2flt::fast_path_correct stdout ----
thread 'num::dec2flt::fast_path_correct' panicked at 'assertion failed: `(left == right)`
  left: `Err(ParseFloatError { kind: Invalid })`,
 right: `Ok(1.448997445238699)`', src/libcore/../libcore/tests/num/dec2flt/mod.rs:97:5
---- num::dec2flt::infinity stdout ----
---- num::dec2flt::infinity stdout ----
thread 'num::dec2flt::infinity' panicked at 'assertion failed: `(left == right)`
  left: `Err(ParseFloatError { kind: Invalid })`,
 right: `Ok(inf)`', src/libcore/../libcore/tests/num/dec2flt/mod.rs:73:5
---- num::dec2flt::large stdout ----
---- num::dec2flt::large stdout ----
thread 'num::dec2flt::large' panicked at 'assertion failed: `(left == right)`
  left: `Err(ParseFloatError { kind: Invalid })`,
 right: `Ok(1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.0)`', src/libcore/../libcore/tests/num/dec2flt/mod.rs:53:5
---- num::dec2flt::ordinary stdout ----
---- num::dec2flt::ordinary stdout ----
thread 'num::dec2flt::ordinary' panicked at 'assertion failed: `(left == right)`
  left: `Err(ParseFloatError { kind: Invalid })`,
 right: `Ok(1.0)`', src/libcore/../libcore/tests/num/dec2flt/mod.rs:27:5
---- num::dec2flt::special_code_paths stdout ----
---- num::dec2flt::special_code_paths stdout ----
thread 'num::dec2flt::special_code_paths' panicked at 'assertion failed: `(left == right)`
  left: `Err(ParseFloatError { kind: Invalid })`,
 right: `Ok(36893488147419103000.0)`', src/libcore/../libcore/tests/num/dec2flt/mod.rs:44:5
---- num::dec2flt::subnormals stdout ----
---- num::dec2flt::subnormals stdout ----
thread 'num::dec2flt::subnormals' panicked at 'assertion failed: `(left == right)`
  left: `Err(ParseFloatError { kind: Invalid })`,
 right: `Ok(0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005)`', src/libcore/../libcore/tests/num/dec2flt/mod.rs:61:5
---- num::dec2flt::zero stdout ----
---- num::dec2flt::zero stdout ----
thread 'num::dec2flt::zero' panicked at 'assertion failed: `(left == right)`
  left: `Err(ParseFloatError { kind: Invalid })`,
 right: `Ok(0.0)`', src/libcore/../libcore/tests/num/dec2flt/mod.rs:81:5

failures:
    num::dec2flt::fast_path_correct
    num::dec2flt::infinity
---
  local time: Sun Jun 21 04:35:00 UTC 2020
  network time: Sun, 21 Jun 2020 04:35:00 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73569/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73569/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (4414) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Aaron1011 commented:
I'm not quite sure what the best way to handle pretty-printing is. There seem to be a few options:

1. Don't insert `None`-delimited groups, and leave `macro_rules!` behavior inconsistent across crates. This seems confusing for users, and has real consequences when writing proc-macros (less span information is available). However, it avoids any breakage.
2. Do what this PR currently does. but special-case `stringify!` to not insert extra spaces for outer (or maybe all?) `Nonterminal`s. This seems like kind of a hack, and will change the behavior of proc-macros that generate calls to `stringify!` with custom tokens.
3. Instead of adding spaces when pretty-printing `None`-delimited groups, just print nothing. I think this may be the best option: since they will be lost on re-parsing, they only serve as an indication that a metavariable replacement occurred (but without providing any additional information). I think it's almost impossible to write any code which actually uses the behavior, and any code which does would be better served by inspecting the `TokenStream` directly. However, this will probably result in some amount of breakage due to crates matching exactly on the pretty-printed output.

ratijas commented:
Hi,

Just throwing my two cents here. I heard about `None`-delimited groups which you are talking about. In fact, I had troubles with them, and found some bugs in rustc AND proc_macro ecosystem: #71422.

Any chances that this PR fixes some of these issues raised there?

Aaron1011 commented:
@ratijas: I don't think this PR will affect #71422, since that issue don't seem to be related to nested `macro_rules!` macros.

Aaron1011 committed:
Don't print additional spaces when pretty-printing NoDelim groups

Aaron1011 commented:
I've switched to removing the extra spaces printed for `None`-delimited groups. This has the nice effect of making re-collecting tokens a no-op in more cases - the pretty-printed output is no longer affected by the replacement of `Nonterminal`s with `None`-delimited groups. We should probably do a Crater run for this.

Aaron1011 commented:
@bors try

bors commented:
:hourglass: Trying commit 2faaf6dc571385cea3f0ba0094beb1c4e7414b28 with merge 4a152b6bdae9e6d4a77446a19ff11d26b8d2b536...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2faaf6dc571385cea3f0ba0094beb1c4e7414b28","merge_sha":"4a152b6bdae9e6d4a77446a19ff11d26b8d2b536"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32594)
Build commit: 4a152b6bdae9e6d4a77446a19ff11d26b8d2b536 (`4a152b6bdae9e6d4a77446a19ff11d26b8d2b536`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32594"},"merge_sha":"4a152b6bdae9e6d4a77446a19ff11d26b8d2b536"} -->

petrochenkov commented:
>To fix this inconsistency, we now insert `None`-delimited groups when 'lowering' a `Nonterminal` `macro_rules!` body, just as we do in `proc_macro_server`.

This is a great catch, I never noticed we forget to wrap tokens into a group during "lowering" nonterminals to HIR.
Serializing them without wrapping is certainly incorrect and may cause mistakes like `$e * 3` where `$e` is `1 * 2` turning into `1 + 2 * 3`.

One problem here is that rustc parser is entirely unready to encounter `None` delimiters and will interpret them as "no parents" anyway - https://github.com/rust-lang/rust/issues/67062, it shouldn't block this fix though.

petrochenkov commented on src/librustc_ast/attr/mod.rs:
```diff
@@ -590,6 +593,11 @@ impl MetaItemKind {
                 tokens.next();
                 MetaItemKind::list_from_tokens(inner_tokens)
             }
+            Some(TokenTree::Delimited(_, token::NoDelim, inner_tokens)) => {
```
Have you seen this case in code? Could you add a test case?
I'm not sure how could `#[attr ⟪ 2 + 2 ⟫]` get through parser.

Aaron1011 replied:
This was needed to get libstd to compile - I'll try to minimize the case I found.

petrochenkov replied:
The added test with `#[doc = $x]` corresponds to the other change above though.

Aaron1011 replied:
You're right - I've removed this match arm.

petrochenkov commented on src/librustc_ast_pretty/pprust.rs:
```diff
@@ -245,7 +245,7 @@ fn token_kind_to_string_ext(tok: &TokenKind, convert_dollar_crate: Option<Span>)
         token::OpenDelim(token::Brace) => "{".to_string(),
         token::CloseDelim(token::Brace) => "}".to_string(),
-        token::OpenDelim(token::NoDelim) | token::CloseDelim(token::NoDelim) => " ".to_string(),
+        token::OpenDelim(token::NoDelim) | token::CloseDelim(token::NoDelim) => "".to_string(),
```
Need a test for a nodelim group between some other tokens.
I'm afraid that `A ⟪B⟫ C` can be printed as `ABC` sometimes turning three tokens into one or two.

Aaron1011 commented:
@petrochenkov: I've added additional test cases

petrochenkov commented:
@bors try

bors commented:
:hourglass: Trying commit b35cb0aeb8e0f095fa9a2506ee438dc4d4c63a07 with merge 433da287daece77315c0cb8c4eb03b037342c115...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b35cb0aeb8e0f095fa9a2506ee438dc4d4c63a07","merge_sha":"433da287daece77315c0cb8c4eb03b037342c115"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32767)
Build commit: 433da287daece77315c0cb8c4eb03b037342c115 (`433da287daece77315c0cb8c4eb03b037342c115`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=32767"},"merge_sha":"433da287daece77315c0cb8c4eb03b037342c115"} -->

Aaron1011 commented:
@craterbot check

pietroalbini commented:
@craterbot abort

pietroalbini commented:
@craterbot check

Aaron1011 commented:
@craterbot check

Aaron1011 commented:
@craterbot abort

Aaron1011 commented:
@craterbot check

Aaron1011 commented:
@craterbot abort

pietroalbini commented:
@craterbot retry

pietroalbini commented:
@craterbot check

pietroalbini commented:
@craterbot abort name=pr-73569

Aaron1011 commented:
There is one legitimate regression: [dmutil](https://github.com/Emoun/dmutil/blob/ac203f595ad2d87a520c5e5e84fd050d0159a9bb/src/eager_macro_rules.rs#L205-L233)

It has some very weird macros:

```rust
	(	@final[
			$(#[$($metas:tt)*])*
			$macro_name:ident $dollar1:tt $id_1:ident
			$({$($rules_grammar:tt)*} => {$($rules_expansion:tt)*})+
		]
	)=>{
		$(#[$($metas)*])*
		macro_rules! $macro_name{
			$(
				// First the eager supporting version
				{
					@eager[$dollar1($dollar1 $id_1:tt)*]
					$($rules_grammar)*
				} => {
					eager_internal!{
						@from_macro[$dollar1($dollar1 $id_1)*]
						$($rules_expansion)*
					}
				};
			)+
			
			$(
				// Then the pure version. We put the pure versions
				// last such that if it contains a '$($all:tt)*' rule,
				// the pure version will not catch an eager call.
				{$($rules_grammar)*} => {$($rules_expansion)*};
			)+
		}
	};
```


As far as I can tell, `dmutil` isn't doing anything wrong. However, we seem to be choking on the `None`-delimited groups when parsing the deserialized `macro_rules!` body (which is why the error only ocurs when running tests). That is, the parser can handle `NtExpr(<inner_tokens>)` or `<inner_tokens>`, but not `NoneGroup(<inner_tokens>)`.

Aaron1011 commented:
Specifically, we're running into problems here:

https://github.com/rust-lang/rust/blob/c977b8775dd72d191ff1d8e8dceaf4b4cd5db86c/src/librustc_expand/mbe/quoted.rs#L112-L118

We're encountering a `None`-delimited group, but this code is only set up to handle groups with normal delimiters.

@petrochenkov: Should I add in enough `None`-delimited group handling to make `dmutil` compile, or should this PR be blocked on more robust handling of `None`-delimited groups?

Aaron1011 committed:
Handle `None`-delimited groups when parsing `macro_rules!` macro

When a `macro_rules!` macro expands to another `macro_rules!` macro, we
may see `None`-delimited groups in odd places when another crate
deserializes the 'inner' macro. This commit 'unwraps' an outer
`None`-delimited group to avoid breaking existing code.

See https://github.com/rust-lang/rust/pull/73569#issuecomment-650860457
for more details.

The proper fix is to handle `None`-delimited groups systematically
throughout the parser, but that will require significant work. In the
meantime, this hack lets us fix important hygiene bugs in macros

Aaron1011 commented on src/test/ui/proc-macro/auxiliary/meta-delim.rs:
```diff
@@ -0,0 +1,12 @@
+    ($dollar_one:tt $foo:ident $my_name:ident) => {
+        #[macro_export]
+        macro_rules! meta_delim {
+            ($dollar_one ($dollar_one $my_name:ident)*) => {
```
:scream:

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=33217&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/33217/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/33217/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux mingw-check
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 63'
Agent machine name: 'fv-az619'
Current agent version: '2.171.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200621.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200621.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.171.1)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/bf975477-a4b5-4bc7-86ca-888422dcdff2.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73569/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73569/merge:refs/remotes/pull/73569/merge
---
 ---> a9ec21d337b3
Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ---> Using cache
 ---> 5ff2c13d8dba
Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            python3 ../x.py doc --stage 0 src/libstd &&            /scripts/validate-toolstate.sh
 ---> 6b931e755c7e
Successfully built 6b931e755c7e
Successfully tagged rust-ci:latest
Built container sha256:6b931e755c7ea1f69816640eb9df74fafd40a545d0e5aa8341d35009dabb0f3c
---
    Checking rustc_span v0.0.0 (/checkout/src/librustc_span)
    Checking rustc_errors v0.0.0 (/checkout/src/librustc_errors)
    Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking tracing v0.1.15
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
    Checking rustc_session v0.0.0 (/checkout/src/librustc_session)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking chalk-solve v0.14.0
---
configure: llvm.assertions      := True
configure: build.locked-deps    := True
configure: rust.dist-src        := False
configure: rust.debug-assertions := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Hugepagesize:       2048 kB
DirectMap4k:      110528 kB
DirectMap2M:     4083712 kB
DirectMap1G:     5242880 kB
+ python3 ../x.py test src/tools/expand-yaml-anchors
Ensuring the YAML anchors in the GitHub Actions config were expanded
Ensuring the YAML anchors in the GitHub Actions config were expanded
Building stage0 tool expand-yaml-anchors (x86_64-unknown-linux-gnu)
---
   Compiling linked-hash-map v0.5.2
   Compiling lazy_static v1.4.0
   Compiling yaml-rust v0.4.3
   Compiling quote v1.0.2
   Compiling thiserror-impl v1.0.5
   Compiling thiserror v1.0.5
   Compiling yaml-merge-keys v0.4.0
   Compiling expand-yaml-anchors v0.1.0 (/checkout/src/tools/expand-yaml-anchors)
Build completed successfully in 0:00:31
+ python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu
    Finished dev [unoptimized] target(s) in 0.18s
Checking rustdoc artifacts (x86_64-unknown-linux-gnu -> i686-pc-windows-gnu)
---
    Checking rustc_span v0.0.0 (/checkout/src/librustc_span)
    Checking rustc_errors v0.0.0 (/checkout/src/librustc_errors)
    Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking tracing v0.1.15
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
    Checking rustc_session v0.0.0 (/checkout/src/librustc_session)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking chalk-solve v0.14.0
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/librustc_expand/mbe/quoted.rs at line 107:
     match tree {
         // `tree` is a `$` token. Look at the next token in `trees`
         tokenstream::TokenTree::Token(Token { kind: token::Dollar, span }) => {
-
             // FIXME: Handle `None`-delimited groups in a more systematic way
             let mut next = outer_trees.next();
Diff in /checkout/src/librustc_expand/mbe/quoted.rs at line 132:
                     // Parse the contents of the sequence itself
                     // Parse the contents of the sequence itself
                     let sequence = parse(tts, expect_matchers, sess, node_id);
                     // Get the Kleene operator and optional separator
-                    let (separator, kleene) = parse_sep_and_kleene_op(&mut trees, span.entire(), sess);
+                    let (separator, kleene) =
+                        parse_sep_and_kleene_op(&mut trees, span.entire(), sess);
                     // Count the number of captured "names" (i.e., named metavars)
                     let name_captures = macro_parser::count_names(&sequence);
                     TokenTree::Sequence(
Diff in /checkout/src/librustc_expand/mbe/quoted.rs at line 160:
 
                 // `tree` is followed by a random token. This is an error.
                 Some(tokenstream::TokenTree::Token(token)) => {
-                    let msg =
-                        format!("expected identifier, found `{}`", pprust::token_to_string(&token),);
+                        "expected identifier, found `{}`",
+                        pprust::token_to_string(&token),
+                    );
+                    );
                     sess.span_diagnostic.span_err(token.span, &msg);
                     TokenTree::MetaVar(token.span, Ident::invalid())
Diff in /checkout/src/librustc_expand/mbe/quoted.rs at line 169:
Diff in /checkout/src/librustc_expand/mbe/quoted.rs at line 169:
                 // There are no more tokens. Just return the `$` we already have.
                 None => TokenTree::token(token::Dollar, span),
-        },
+        }
 
 
         // `tree` is an arbitrary token. Keep it.
         tokenstream::TokenTree::Token(token) => TokenTree::Token(token),
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustc_expand/mbe/quoted.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:42
== clock drift check ==
  local time: Mon Jun 29 14:21:40 UTC 2020
  network time: Mon, 29 Jun 2020 14:21:40 GMT
  network time: Mon, 29 Jun 2020 14:21:40 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73569/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73569/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (5845) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Aaron1011 commented:
@petrochenkov: I've added a workaround to get `dmutil` compiling. This PR should be ready to merge.

petrochenkov commented:
@Aaron1011 
Do I understand correctly that in the regression case `NtTT` (`$dollar_one`) is encountered during AST lowering and it gets wrapped into a `None` group?

In that case it's not correct, `NtTT` must not be wrapped into a group, it must be passed as is.
I just assumed that `NtTT` can never be passed to `nt_to_tokenstream`, neither in proc_macro_server, nor in AST lowering, just because it's substituted early.
But apparently that's not the case for generated `macro_rules`.

Aaron1011 commented:
@petrochenkov: The `NtTT` doesn't get wrapped - whatever occurs after the dollar sign (an ident or delimited group) does.

petrochenkov commented:
Ah, ok, so we get `$ ⟪var⟫` or `$⟪(tokens)⟫*` and `macro_rules` can't parse it because it doesn't do the same automatic flattening of `NoDelim` groups as `TokenCursor::next` (IIRC) does, and sees the original token stream.

petrochenkov commented:
Seems ok to patch `macro_rules` like this, it's basically the same thing as the attribute changes from the earlier commits, attributes see the original token stream without `NoDelim` group flattening as well.

@bors r+
(The PR could use some squashing to eliminate fixup commits though.)

bors commented:
:pushpin: Commit 4cd31965c91c7eefff880075426e35846e2ec22e has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 4cd31965c91c7eefff880075426e35846e2ec22e -->
<!-- homu: {"type":"Approved","sha":"4cd31965c91c7eefff880075426e35846e2ec22e","approver":"petrochenkov"} -->

Aaron1011 commented:
Squashed into more meaningful commits

@bors r=petrochenkov

bors commented:
:pushpin: Commit 1ded7a581589b65be35724ea00bb5cec4e13c673 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 1ded7a581589b65be35724ea00bb5cec4e13c673 -->
<!-- homu: {"type":"Approved","sha":"1ded7a581589b65be35724ea00bb5cec4e13c673","approver":"petrochenkov"} -->
